### PR TITLE
Improve Slack error handling

### PR DESCRIPTION
**Pull request by vadimkerr**
# What this PR does

- Rename `SlackClientWithErrorHandling` to just `SlackClient`
- Add more error classes + improve the way errors are raised based on the Slack error code
- Add API call retries on Slack server errors (e.g. when Slack returns `5xx` errors)
- Refactor some methods working with Slack API + add tests

## Which issue(s) this PR fixes

- https://github.com/grafana/oncall-private/issues/1837
- https://github.com/grafana/oncall-private/issues/1840
- https://github.com/grafana/oncall-private/issues/1842

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/apps/slack/alert_group_slack_service.py [hunk 1]
```diff
@@ -42,28 +45,22 @@ def update_alert_group_slack_message(self, alert_group: "AlertGroup") -> None:
...
-                logger.info(f"Skip updating slack message for alert_group {alert_group.pk} due message_not_found")
-            elif e.response["error"] == "is_inactive":  # deleted channel error
-                logger.info(f"Skip updating slack message for alert_group {alert_group.pk} due to is_inactive")
-            elif e.response["error"] == "account_inactive":
-                logger.info(f"Skip updating slack message for alert_group {alert_group.pk} due to account_inactive")
-            elif e.response["error"] == "channel_not_found":
-                logger.info(f"Skip updating slack message for alert_group {alert_group.pk} due to channel_not_found")
-            else:
-                raise e
-        logger.info(f"Finished _update_slack_message for alert_group {alert_group.pk}")
```
Comment (vadimkerr) [inline engine/apps/slack/alert_group_slack_service.py:66]: Removing logging here as errors are already logged in [`SlackClient.api_call`](https://github.com/grafana/oncall/blob/091792cdad2fd8235ad3298341f5e31ecd2c9d1c/engine/apps/slack/client.py#L128-L138)

File: engine/apps/slack/client.py [hunk 2]
```diff
@@ -1,36 +1,65 @@
...
+    from apps.slack.models import SlackTeamIdentity
 
 logger = logging.getLogger(__name__)
 
 
-class SlackAPIException(Exception):
-    def __init__(self, msg: str, response: SlackResponse):
-        super().__init__(msg)
-        self.response = response
+class SlackServerErrorRetryHandler(RetryHandler):
```
Comment (vadimkerr) [inline engine/apps/slack/client.py:25]: ⬆️ One of the main changes in the PR

File: engine/apps/slack/client.py [hunk 3]
```diff
@@ -84,49 +113,61 @@ def paginated_api_call_with_ratelimit(
...
             rate_limited = True
 
         return cumulative_response, cursor, rate_limited
 
-    def api_call(self, *args, **kwargs):
-        try:
-            response = super(SlackClientWithErrorHandling, self).api_call(*args, **kwargs)
-        except SlackApiError as err:
-            response = err.response
+    def api_call(self, *args, **kwargs) -> SlackResponse:
```
Comment (vadimkerr) [inline engine/apps/slack/client.py:121]: ⬆️ One of the main changes in the PR

File: engine/apps/slack/errors.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+from apps.slack.constants import SLACK_RATE_LIMIT_DELAY
+
+
+class UnexpectedResponse(typing.TypedDict):
+    status: int
+    headers: dict[str, typing.Any]
+    body: str
+
+
+class SlackAPIError(Exception):
```
Comment (vadimkerr) [inline engine/apps/slack/errors.py:14]: ⬆️ One of the main changes in the PR

File: engine/apps/slack/models/slack_usergroup.py [hunk 5]
```diff
@@ -123,59 +124,24 @@ def get_users_from_members_for_organization(self, organization):
...
+        sc = SlackClient(slack_team_identity)
+        usergroups = sc.usergroups_list()["usergroups"]
+
+        usergroup = [ug for ug in usergroups if ug["id"] == slack_id][0]
         try:
-            usergroups_list = sc.usergroups_list()
-        except SlackAPIException as e:
-            if e.response["error"] == "not_allowed_token_type":
-                # Trying same request with access token. It is required due to migration to granular permissions
-                # and can be removed after clients reinstall their bots
```
Comment (vadimkerr) [inline engine/apps/slack/models/slack_usergroup.py:134]: > can be removed after clients reinstall their bots

The switch to granular permissions happened a long time ago so I think it's safe to remove this

**Review by matiasb — APPROVED** [review]
LGTM, a couple of minor questions.

File: engine/apps/slack/scenarios/distribute_alerts.py [hunk 6]
```diff
@@ -917,56 +892,22 @@ def process_signal(self, log_record: AlertGroupLogRecord) -> None:
...
-                        f"{e}"
-                    )
-                elif e.response["error"] == "is_archived":
-                    logger.warning(
-                        f"Unable to delete messages in slack. Message ts: {message_ts}"
-                        f"Slack team identity pk: {self.slack_team_identity.pk}.\n"
-                        f"Reason: 'is_archived'"
-                        f"{e}"
-                    )
-                elif e.response["error"] == "cant_delete_message":
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/slack/scenarios/distribute_alerts.py:948]: Yes, I believe we shouldn't see any `cant_delete_message` errors after switching to granular permissions

File: engine/apps/slack/client.py [hunk 7]
```diff
@@ -84,49 +113,61 @@ def paginated_api_call_with_ratelimit(
...
+    @staticmethod
+    def _get_error_class(response: UnexpectedResponse | SlackResponse) -> typing.Type[SlackAPIError]:
+        """Get an appropriate error class for the response"""
+
+        if isinstance(response, dict):  # UnexpectedResponse
+            return SlackAPIServerError
+
+        for error_class in SlackAPIError.__subclasses__():
+            if response["error"] in error_class.errors:
+                return error_class
```
[the remark this answers is not in the document]
Comment (vadimkerr) [reply engine/apps/slack/client.py]: Good call! I agree there's no need to make it more dynamic than it needs to be: https://github.com/grafana/oncall/pull/3000/commits/5b1be852a16c107d7a62cbce4b35ee3f8f67718e.

**Review by joeyorlando — COMMENTED** [review]
awesome cleanup/deduplication of the error handling logic 🔥

**Pull request merged**